### ISSUE TITLE
Meteor build client-only

### ISSUE THREAD

**Issue by raix** [MeteorCommunity/discussions#6]
Fabian Vogelsteller (@frozeman):
I was requesting this feature a while back, but with the new platform system of Meteor it is very possible now.

I would like to have a client-only platform, where it creates a bundle with only the client side files and an "index.html" as start file.
This would allow to host the client side part of a meteor app on any static server, as well as use it in a native app bundler like tideSDK, or Sentenza Desktop.

I would like to see the following command possible:

``` bash
 $ meteor add platform client-only
```

Which then produces a bundles tar ball containing only the index.html, bundle.js and bundle.css.

This would make meteor also immediately available for people who want to use it as fronted library instead of angularjs (growing user base ;)
## 

Zoltan Olah (@zol):
The outstanding issue is that by default the DDP package assumes there is a server to connect back to. Would MDG be open to changing this behavior - perhaps passing something in via **meteor_runtime_config** that disables the autoconnect?
## 

Ref: 
https://github.com/SpaceCapsule/packmeteor/issues/32
https://groups.google.com/forum/#!topic/meteor-core/Pa2Zi0rA7Sg

**Comment by raix** [comment]
@zol maybe the connection could be optional - eg. if bundling for cdn's we might still want the connection but not the hot code push.

``` bash
# No connection - client-only
$ meteor add platform client-only

# For CDN deployment
$ meteor add platform client-only -h "http://my.app.com"

# Or
$ meteor build client-only
```

**Comment by IstoraMandiri** [comment]
:+1:

**Comment by PEM--** [comment]
:+1: With the appearance of DDP on other server stacks, devs using already made server (Wordpress, Prestashop, ...) could switch gradually to Meteor with confidence. Actually, porting DDP to the major other server stack could also really help breakthrough in adoption.

**Comment by grigio** [comment]
:+1:  I suggest to specify just the `hostname.domain` and select "http/ws" or "https/wss" automatically

**Comment by Slava** [comment]
https://groups.google.com/forum/#!topic/meteor-core/Pa2Zi0rA7Sg

**Comment by raix** [comment]
Thanks @Slava its linked under "references" in the top description

**Comment by gadicc** [comment]
Been a while since I added anything to Meteorpedia... added those initial instructions plus a ref to the intended pattern and this thread at http://www.meteorpedia.com/read/Building_for_the_client_only

**Comment by raix** [comment]
Just a note - would something like `$ meteor build client-only <output path>` _(with --server / --directory / --debug as options)_ be sufficient?

`--server` will determine if ddp should be present eg. for cdn?

**Comment by gadicc** [comment]
I quite like that.  Short and sweet.  Although I admit I've never had the need for this before, so could be missing something.  I'd also suggest `--ddp=` which is maybe a little clearer on what server we're referring to.

**Comment by frozeman** [comment]
Hi,
thanks @raix for continuing the discussion. I personally think now that the `client-only` might be to general and we should better have something like `meteor build browser` or `browser-only`.

Currently if you look into the `.meteor/platforms` file you will find `server` and `browser` already exist.
though the `browser-only` or even `server-only` should be available and build a proper index.html.

separating the two this way would help to host meteor on other server infrastructures, or of example to use the meteor ddp connection only for specific actions (even by multiple and native apps)

I think this, with the right showcases and tutorials would do a great deal for meteor adoption!

**Comment by tonyaom** [comment]
+1 billion if possible

**Comment by ctzurcanu** [comment]
+1

**Comment by loredanacirstea** [comment]
+1

**Comment by mitar** [comment]
I would love that as well. I opened a discussion about this [on the mailing list](https://groups.google.com/forum/#!msg/meteor-talk/EvY9uH-ilfs/rOSz61g0Y_0J) some time ago. I think this would be great with combination of WebRTC for DDP protocol so one could use it to talk between browsers directly. A simple list of peers served together with static index.html could make the trick. (That list of peers could be maintained with any other server-side software, much simpler than Meteor or node.js itself.)

**Comment by mitar** [comment]
Related: https://github.com/meteor/meteor/issues/1223

**Comment by merunga** [comment]
I found this [Standalone Blaze Generator](https://github.com/meteor/standalone-blaze-generator), and tweaked it to bundle minimongo, blaze and session packages in a single js file for the browser, with no sockjs connection error or the need for `Meteor.disconnect()`.

Haven't tested it at all beyond this [example](http://merunga.github.io/meteor-client-only-generator/) on gh-pages.
- [repo](https://github.com/merunga/meteor-client-only-generator)
- [example code](https://github.com/merunga/meteor-client-only-generator/blob/master/index.html)

Is this useful for anyone? Make any sense?

**Comment by frozeman** [comment]
Looks very good, thanks. Any ideas how to precompile all the templates? Personally i love the meteor command line tool, so i don't have to link, minify, concatenate and require.js all the files ... again.

I just found ractive.js, which looks very similar to Spacebars, though i love the reactive helpers and Tracker better.

**Comment by sferoze** [comment]
Hi,

I am making a Cordova app using Meteor to communicate with a bluetooth device. We do not need a server, or any internet connection. 

I currently plan on just using Meteor.disconnect 

but that means I would still be shipping connection code even though i am not using it. 

I want to take this out. 

How can I do this before submitting to the app store? 

Thanks in advance!

**Comment by zimme** [comment]
I think one approach is to see which packages the `meteor-platorm` package depends on.

Then instead of using `meteor-platform` you use all the separate packages instead, removing the one handling connections.

But this might not work because a lot of packages actually depend in connection logic as you can
manually specify a connection per collection and such.

I'm talking about core packages.

**Comment by mitar** [comment]
A bit tangential, but I made this package once when I needed to compile custom Meteor package and then inject the resulted JS into the client only: https://github.com/peerlibrary/meteor-package-builder

Maybe you could use to build the client JS and then use only that.

**Comment by Maxim-Filimonov** [comment]
My 5 cents...
Not sure why MDG is so resistant to make this happen. I understand that current focus is new fancy companies who can write greenfield projects BUT meteor is gonna push itself into the same trap which angular fall into. You either write an angular app THE angular way or you hack around by passing json into ng-init. Which made angular VERY hard sell for any existing interfaces. 

Our use case for client side only build:
We have couple of areas in our existing UI which we want to use meteor for, we don't have luxury to just throw away all existing code and rewrite everything in meteor. We have meteor server running connected to a mongo database, so we thought why not to use meteor for those UI pieces where we want to have reactivity and slowly adapt meteor in all other parts going forward.

**Comment by mitar** [comment]
Some of this would be interesting to integrate with Meteor:
- https://github.com/HelloZeroNet/ZeroNet
- https://github.com/feross/webtorrent
- https://github.com/substack/hyperboot

**Comment by frozeman** [comment]
**I finally did it myself :)**
https://github.com/frozeman/meteor-build-client

**Comment by sferoze** [comment]
@frozeman Can this work to build a cordova mobile app that is for the client only? It looks like it is only for web currently.

**Comment by Maxim-Filimonov** [comment]
@frozeman Awesome work! :+1:

**Comment by frozeman** [comment]
@sferoze isn't the meteor native cordova deploy not enough?
i'm currently not planning in adding any cordova feature, as i wanted only to get the minimalistic files to run in everywhere (or even by just opening the index.html)

**Comment by sferoze** [comment]
@frozeman Well when you build your app for Cordova you are expected to input a mobile server address. This is because the DDP package is still part of the mobile app. Currently I have in the client code Meteor.disconnect() so the app does not use the DDP package. But this is still not optimal.

**Comment by frozeman** [comment]
The client side part generated with `meteor-build-client` also uses `Meteor.disconnect()` to prevent any server connection (which would lead to an endless reload cycle).

Its not ideal, but it works.

**Comment by rurri** [comment]
This is a great idea.

+1

**Comment by sdomagala** [comment]
Any idea if you can make some kind of routing for client-side only?

**Comment by frozeman** [comment]
Use iron-router or flow-router.

**Comment by sdomagala** [comment]
Obviously, I forgot to say I don't wanna use meteor for it, just blaze for static content

**Comment by frozeman** [comment]
Look at my tool: https://www.npmjs.com/package/meteor-build-client

This lets you develop you app with meteor, and get on the end static files with index.html ;)

**Comment by vyky** [comment]
Hi @sferoze what you put on --server flag when you build meteor cordova client only?
